<a href="https://colab.research.google.com/github/CodeHunterOfficial/NLP-2024-2025/blob/main/Lecture_3_2_%D0%9F%D0%BE%D0%B8%D1%81%D0%BA_%D1%81%D0%BC%D1%8B%D1%81%D0%BB%D0%B0_%D1%81%D0%BB%D0%BE%D0%B2_%D0%BF%D0%BE_%D0%B8%D1%85_%D1%87%D0%B0%D1%81%D1%82%D0%BE%D1%82%D0%BD%D0%BE%D1%81%D1%82%D1%8F%D0%BC_%D1%81%D0%B5%D0%BC%D0%B0%D0%BD%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B9_%D0%B0%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Поиск смысла слов по их частотностям: семантический анализ

Введение в семантический анализ и векторное представление слов является основой для разработки современных систем обработки естественного языка (NLP). Эта лекция охватывает несколько ключевых аспектов семантического анализа, включая создание векторов тем, семантический поиск, масштабируемый анализ для больших корпусов текстов и использование семантических компонентов в NLP. Особое внимание будет уделено ориентации в векторных пространствах высокой размерности.

#### 1. Создание векторов тем с помощью анализа семантики (смысла)

**1.1 Основы семантического анализа**

Семантический анализ направлен на выявление смысла слов и предложений путем изучения их контекста. Одним из основных методов является векторное представление слов, также известное как word embeddings. Такие методы, как Word2Vec, GloVe и FastText, используют статистические модели для представления слов в виде векторов в высокоразмерном пространстве, где семантически схожие слова располагаются ближе друг к другу.

**1.2 Модели для создания векторов тем**

1. **Word2Vec**: Использует модели CBOW и Skip-Gram для обучения векторных представлений слов на больших корпусах текста. CBOW предсказывает слово по контексту, а Skip-Gram делает обратное.
2. **GloVe (Global Vectors for Word Representation)**: Основан на матрицах соотношений слов и глобальной статистике корпуса.
3. **FastText**: Улучшение Word2Vec, учитывающее морфологическую структуру слов за счет использования субслов.



Рассмотрим каждую модели

Word2Vec — алгоритм, предлагающий два основных подхода: CBOW (Continuous Bag of Words) и Skip-Gram. Мы рассмотрим оба подхода с соответствующими формулами и примерами.

**1.1 CBOW (Continuous Bag of Words)**

CBOW предсказывает текущее слово по его контексту.

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \log P(w_t | w_{t-i}, \ldots, w_{t+i}) $$

**Вероятность**:
$$ P(w_t | w_{t-i}, \ldots, w_{t+i}) = \frac{\exp(v_{w_t} \cdot h)}{\sum_{w' \in V} \exp(v_{w'} \cdot h)} $$

где $ h $ — средний вектор контекстных слов:
$$ h = \frac{1}{2i} \sum_{j=-i, j \neq 0}^{i} v_{w_{t+j}} $$

**Пример**:
Для предложения "I love natural language processing", если текущее слово —

**Пример предложения**: "I love natural language processing".

**Контекстное окно**: 2 (по 2 слова слева и справа от целевого слова).

**Целевое слово**: "language".

**Контекстные слова**: ["I", "love", "natural", "processing"].

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \log P(w_t | w_{t-i}, \ldots, w_{t+i}) $$

Для нашего примера:

$$ J(\theta) = \log P(language | I, love, natural, processing) $$

**Вероятность**:
$$ P(w_t | w_{t-i}, \ldots, w_{t+i}) = \frac{\exp(v_{w_t} \cdot h)}{\sum_{w' \in V} \exp(v_{w'} \cdot h)} $$

где $ h $ — средний вектор контекстных слов:

$$h = \frac{1}{4} (v_{I} + v_{love} + v_{natural} + v_{processing}) $$

**Шаг 1: Вычисление среднего вектора контекста**

Пусть векторы слов имеют размерность 3 для простоты. Допустим, что векторы слов такие:

- $ v_{I} = [0.2, 0.1, -0.1] $
- $ v_{love} = [0.4, 0.3, 0.2] $
- $ v_{natural} = [0.3, 0.4, 0.5] $
- $ v_{processing} = [0.5, 0.6, 0.7] $

Тогда средний вектор контекста:

$$ h = \frac{1}{4} ([0.2, 0.1, -0.1] + [0.4, 0.3, 0.2] + [0.3, 0.4, 0.5] + [0.5, 0.6, 0.7]) $$
$$ h = \frac{1}{4} ([1.4, 1.4, 1.3]) $$
$$ h = [0.35, 0.35, 0.325] $$

**Шаг 2: Вычисление вероятности целевого слова**

Пусть вектор целевого слова $ v_{language} = [0.6, 0.7, 0.8] $.

Скалярное произведение:

$$ v_{language} \cdot h = [0.6, 0.7, 0.8] \cdot [0.35, 0.35, 0.325] $$
$$ v_{language} \cdot h = 0.6 \cdot 0.35 + 0.7 \cdot 0.35 + 0.8 \cdot 0.325 $$
$$ v_{language} \cdot h = 0.21 + 0.245 + 0.26 $$
$$ v_{language} \cdot h = 0.715 $$

Теперь вероятность:

$$ P(language | I, love, natural, processing) = \frac{\exp(0.715)}{\sum_{w' \in V} \exp(v_{w'} \cdot h)} $$

Для простоты предположим, что сумма экспонент для всех слов в словаре равна $ Z $:

$$ P(language | I, love, natural, processing) = \frac{\exp(0.715)}{Z} $$

Целевая функция:

$$ J(\theta) = \log P(language | I, love, natural, processing) = 0.715 - \log Z $$



In [ ]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Пример предложения
sentences = ["I love natural language processing"]

# Параметры модели
window_size = 2
embedding_dim = 3
epochs = 100

# Токенизация и создание словаря
tokenizer = Tokenizer()
tokenizer.fit_on_texts(sentences)
word_index = tokenizer.word_index
vocab_size = len(word_index) + 1
sequences = tokenizer.texts_to_sequences(sentences)

# Функция для создания данных для CBOW
def generate_cbow_data(sequences, window_size):
    data = []
    labels = []
    for sequence in sequences:
        for i, word_id in enumerate(sequence):
            context = []
            for j in range(-window_size, window_size + 1):
                if j != 0 and 0 <= i + j < len(sequence):
                    context.append(sequence[i + j])
            if len(context) == 2 * window_size:
                data.append(context)
                labels.append(word_id)
    return np.array(data), np.array(labels)

# Создание данных для обучения
data, labels = generate_cbow_data(sequences, window_size)

# Модель CBOW
input_layer = tf.keras.layers.Input(shape=(2 * window_size,), name='input')
embedding_layer = tf.keras.layers.Embedding(vocab_size, embedding_dim, input_length=2 * window_size, name='embedding')(input_layer)
context_layer = tf.keras.layers.Lambda(lambda x: tf.reduce_mean(x, axis=1), name='context')(embedding_layer)
output_layer = tf.keras.layers.Dense(vocab_size, activation='softmax', name='output')(context_layer)

model = tf.keras.Model(inputs=input_layer, outputs=output_layer)
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam')
model.summary()

# Обучение модели
model.fit(data, labels, epochs=epochs, verbose=2)

# Получение векторов слов
word_embeddings = model.get_layer('embedding').get_weights()[0]
print("Word Embeddings:", word_embeddings)

Model: "model_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input (InputLayer)          [(None, 4)]               0         
                                                                 
 embedding (Embedding)       (None, 4, 3)              18        
                                                                 
 context (Lambda)            (None, 3)                 0         
                                                                 
 output (Dense)              (None, 6)                 24        
                                                                 
Total params: 42 (168.00 Byte)
Trainable params: 42 (168.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
1/1 - 0s - loss: 1.7819 - 479ms/epoch - 479ms/step
Epoch 2/100
1/1 - 0s - loss: 1.7793 - 8ms/epoch - 8ms/step
Epoch 3/100
1/1 - 0s - loss: 1.77














**1.2 Skip-Gram**

Skip-Gram предсказывает контекстные слова по текущему слову.

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \sum_{-i \leq j \leq i, j \neq 0} \log P(w_{t+j} | w_t) $$

**Вероятность**:
$$ P(w_{t+j} | w_t) = \frac{\exp(v_{w_{t+j}} \cdot v_{w_t})}{\sum_{w' \in V} \exp(v_{w'} \cdot v_{w_t})} $$

**Пример**:



**Пример предложения**: "I love natural language processing".

**Контекстное окно**: 2 (по 2 слова слева и справа от целевого слова).

**Целевое слово**: "language".

**Контекстные слова**: ["I", "love", "natural", "processing"].

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \sum_{-i \leq j \leq i, j \neq 0} \log P(w_{t+j} | w_t) $$

Для нашего примера:

$$ J(\theta) = \log P(I | language) + \log P(love | language) + \log P(natural | language) + \log P(processing | language) $$

**Вероятность**:
$$ P(w_{t+j} | w_t) = \frac{\exp(v_{w_{t+j}} \cdot v_{w_t})}{\sum_{w' \in V} \exp(v_{w'} \cdot v_{w_t})} $$

**Шаг 1: Вычисление вероятности для каждого контекстного слова**

Пусть вектор целевого слова $ v_{language} = [0.6, 0.7, 0.8] $.

Пусть векторы контекстных слов такие:

- $ v_{I} = [0.2, 0.1, -0.1] $
- $ v_{love} = [0.4, 0.3, 0.2] $
- $ v_{natural} = [0.3, 0.4, 0.5] $
- $ v_{processing} = [0.5, 0.6, 0.7] $

**Вероятность для слова "I"**:

Скалярное произведение:

$$ v_{I} \cdot v_{language} = [0.2, 0.1, -0.1] \cdot [0.6, 0.7, 0.8] $$
$$ v_{I} \cdot v_{language} = 0.2 \cdot 0.6 + 0.1 \cdot 0.7 + (-0.1) \cdot 0.8 $$
$$ v_{I} \cdot v_{language} = 0.12 + 0.07 - 0.08 $$
$$ v_{I} \cdot v_{language} = 0.11 $$

Теперь вероятность:

$$ P(I | language) = \frac{\exp(0.11)}{\sum_{w' \in V} \exp(v_{w'} \cdot v_{language})} $$

Для простоты предположим, что сумма экспонент для всех слов в словаре равна $ Z $:

$$ P(I | language) = \frac{\exp(0.11)}{Z} $$

Аналогично рассчитываются вероятности для остальных контекстных слов.

**Шаг 2: Целевая функция**

Целевая функция:

$$ J(\theta) = \log P(I | language) + \log P(love | language) + \log P(natural | language) + \log P(processing | language) $$

После вычисления всех скалярных произведений и подстановки значений вероятностей, целевая функция примет вид:

$$ J(\theta) = 0.11 - \log Z + 0.26 - \log Z + 0.37 - \log Z + 0.71 - \log Z $$






In [ ]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Пример предложения
sentences = ["I love natural language processing"]

# Параметры модели
window_size = 2
embedding_dim = 3
epochs = 100

# Токенизация и создание словаря
tokenizer = Tokenizer()
tokenizer.fit_on_texts(sentences)
word_index = tokenizer.word_index
vocab_size = len(word_index) + 1
sequences = tokenizer.texts_to_sequences(sentences)

# Функция для создания данных для Skip-Gram
def generate_skipgram_data(sequences, window_size):
    data = []
    labels = []
    for sequence in sequences:
        for i, word_id in enumerate(sequence):
            for j in range(-window_size, window_size + 1):
                if j != 0 and 0 <= i + j < len(sequence):
                    context_word_id = sequence[i + j]
                    data.append(word_id)
                    labels.append(context_word_id)
    return np.array(data), np.array(labels)

# Создание данных для обучения
data, labels = generate_skipgram_data(sequences, window_size)

# Модель Skip-Gram
input_layer = tf.keras.layers.Input(shape=(1,))
embedding_layer = tf.keras.layers.Embedding(vocab_size, embedding_dim, input_length=1)(input_layer)
embedding_layer = tf.keras.layers.Reshape((embedding_dim,))(embedding_layer)
output_layer = tf.keras.layers.Dense(vocab_size, activation='softmax')(embedding_layer)

model = tf.keras.Model(inputs=input_layer, outputs=output_layer)
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam')
model.summary()

# Обучение модели
model.fit(data, labels, epochs=epochs, verbose=2)

# Получение векторов слов
word_embeddings = model.get_layer('embedding').get_weights()[0]
print("Word Embeddings:", word_embeddings)

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 1)]               0         
                                                                 
 embedding (Embedding)       (None, 1, 3)              18        
                                                                 
 reshape (Reshape)           (None, 3)                 0         
                                                                 
 dense (Dense)               (None, 6)                 24        
                                                                 
Total params: 42 (168.00 Byte)
Trainable params: 42 (168.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
1/1 - 1s - loss: 1.7868 - 586ms/epoch - 586ms/step
Epoch 2/100
1/1 - 0s - loss: 1.7859 - 8ms/epoch - 8ms/step
Epoch 3/100
1/1 - 0s - loss: 1.7850

**FastText**

FastText — это модель для обучения векторным представлениям слов и текстов, разработанная и поддерживаемая Facebook AI Research (FAIR). Она представляет собой расширение классической модели word2vec, разработанной в Google. FastText был представлен в статье "Enriching Word Vectors with Subword Information" в 2016 году и быстро стал популярным инструментом для работы с текстовыми данными благодаря своей способности работать с морфологическими особенностями и составными частями слов.

#### Архитектура FastText

FastText использует два основных компонента: модель для обучения векторных представлений слов и метод для генерации и использования подсловных (subword) представлений.

##### 1. Модель векторных представлений слов

FastText строит векторные представления слов, учитывая как сами слова, так и их составляющие (подслова). Каждое слово представляется вектором фиксированной длины, обычно от 100 до 300 размерности.

Для обучения векторов используется skip-gram подход, аналогичный модели word2vec. Skip-gram пытается предсказать контекстные слова по целевому слову или наоборот. Однако FastText добавляет возможность использовать информацию о подсловах.

##### 2. Использование подсловных (subword) представлений

Основная особенность FastText — это способность работать с подсловными (subword) представлениями. Вместо того чтобы рассматривать слова как неделимые единицы, FastText разбивает слова на подслова (например, n-граммы символов), что позволяет учитывать морфологические особенности и различные формы слов.

Конкретно, FastText использует метод буквенных n-грамм для создания векторов подслов. Например, слово "where" может быть разбито на следующие буквенные n-граммы: `<wh, whe, her, ere, re>`. Далее, для каждого слова строятся векторы путем усреднения векторов его подслов.

#### Обучение FastText

Для обучения FastText используется метод градиентного спуска с мини-батчами. Целевая функция FastText включает в себя две основные составляющие:

##### 1. Softmax функция для предсказания вероятности слова в контексте:
$$ P(w_o | w_i) = \frac{\exp(\mathbf{u}_{w_o}^\top \mathbf{v}_{w_i})}{\sum_{j=1}^{|V|} \exp(\mathbf{u}_j^\top \mathbf{v}_{w_i})} $$

где:
- $ w_i $ — целевое слово,
- $ w_o $ — слово из контекста,
- $ \mathbf{v}_{w_i} $ — вектор слова $ w_i $,
- $ \mathbf{u}_{w_o} $ — вектор слова $ w_o $,
- $ |V| $ — размер словаря.

##### 2. Negative sampling для улучшения обучения:
$$ \mathcal{L}_{\text{neg}} = - \log \sigma(\mathbf{u}_{w_o}^\top \mathbf{v}_{w_i}) - \sum_{k=1}^K \mathbb{E}_{w_k \sim P_{\text{noise}}(w)}[\log \sigma(-\mathbf{u}_{w_k}^\top \mathbf{v}_{w_i})] $$

где $ \sigma(x) = \frac{1}{1 + \exp(-x)} $, $ P_{\text{noise}}(w) $ — распределение шумовых слов, $ K $ — количество шумовых слов.

#### Применение FastText

FastText находит применение в различных задачах обработки текстов, таких как классификация текстов, кластеризация, поиск похожих документов и других задачах, где важна работа с текстовыми данными и учет морфологии слов.



**Пример.**

Теперь, давайте вычислим Softmax Loss и Negative Sampling Loss для заданной пары слов.

Предположим, у нас есть следующие данные:
- Целевое слово $ w_i = \text{"apple"} $
- Контекстное слово $ w_o = \text{"juicy"} $
- Размер словаря $ |V| = 10000 $
- Векторные представления слов:
  - $ \mathbf{v}_{\text{"apple"}} = [0.2, -0.3, 0.5, \ldots] $ (размерность 300)
  - $ \mathbf{u}_{\text{"juicy"}} = [0.1, 0.4, -0.2, \ldots] $ (размерность 300)

#### Вычисление Softmax Loss

Softmax Loss для предсказания контекстного слова "juicy" по целевому слову "apple" вычисляется как:
$$ \mathcal{L}_{\text{softmax}} = -\log P(\text{"juicy"} | \text{"apple"}) $$
$$ P(\text{"juicy"} | \text{"apple"}) = \frac{\exp(\mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}})}{\sum_{j=1}^{|V|} \exp(\mathbf{u}_j^\top \mathbf{v}_{\text{"apple"}})} $$

Подставим значения:
$$ \mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}} = 0.1 \cdot 0.2 + 0.4 \cdot (-0.3) + (-0.2) \cdot 0.5 + \ldots $$
$$ \text{(вычисляем скалярное произведение векторов)} $$

Предположим, что получили $ \mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}} = 0.8 $.

Тогда:
$$ P(\text{"juicy"} | \text{"apple"}) = \frac{\exp(0.8)}{\sum_{j=1}^{10000} \exp(\mathbf{u}_j^\top \mathbf{v}_{\text{"apple"}})} $$

Для вычисления Softmax Loss используется логарифм:
$$ \mathcal{L}_{\text{softmax}} = -\log \left( \frac{\exp(0.8)}{\sum_{j=1}^{10000} \exp(\mathbf{u}_j^\top \mathbf{v}_{\text{"apple"}})} \right) $$

#### Вычисление Negative Sampling Loss

Negative Sampling Loss для той же пары слов выглядит следующим образом:
$$ \mathcal{L}_{\text{neg}} = -\log \sigma(\mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}}) - \sum_{k=1}^K \mathbb{E}_{w_k \sim P_{\text{noise}}(w)}[\log \sigma(-\mathbf{u}_{w_k}^\top \mathbf{v}_{\text{"apple"}})] $$

Для примера, допустим $ K = 5 $ (5 шумовых слов):
$$ \sigma(x) = \frac{1}{1 + \exp(-x)} $$

Тогда первое слагаемое:
$$ \sigma(\mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}}) = \sigma(0.8) = \frac{1}{1 + \exp(-0.8)} $$

А второе слагаемое:
$$ \sum_{k=1}^5 \mathbb{E}_{w_k \sim P_{\text{noise}}(w)}[\log \sigma(-\mathbf{u}_{w_k}^\top \mathbf{v}_{\text{"apple"}})] $$

Здесь $ \mathbb{E}_{w_k \sim P_{\text{noise}}(w)} $ представляет математическое ожидание по распределению шумовых слов.

Таким образом, Softmax Loss и Negative Sampling Loss вычисляются на основе векторных представлений слов и используются для обучения модели FastText на больших текстовых корпусах.






In [ ]:
import numpy as np

# Заданные векторные представления слов
v_apple = np.array([0.2, -0.3, 0.5, 0.1])  # вектор представления слова "apple"
u_juicy = np.array([0.1, 0.4, -0.2, -0.7])  # вектор представления слова "juicy"

# Размер словаря
V_size = 10000

# Функция для вычисления Softmax Loss
def softmax_loss(u, v, V_size):
    dot_product = np.dot(u, v)
    exp_dot_product = np.exp(dot_product)

    softmax = exp_dot_product / np.sum(np.exp(np.dot(u, v)))
    loss = -np.log(softmax)

    return loss

# Функция для вычисления Negative Sampling Loss
def negative_sampling_loss(u, v, K, V_size):
    dot_product = np.dot(u, v)
    sigmoid_dot_product = 1 / (1 + np.exp(-dot_product))

    # Генерация шумовых слов
    noise_words = np.random.choice(V_size, K, replace=False)

    # Вычисление второго слагаемого
    neg_loss = 0
    for noise_word in noise_words:
        # Получение вектора шумового слова
        v_k = np.random.uniform(-1, 1, size=len(u))
        neg_dot_product = np.dot(u, v_k)
        neg_loss += np.log(1 + np.exp(neg_dot_product))

    loss = -np.log(sigmoid_dot_product) + neg_loss

    return loss

# Пример использования функций
softmax_loss_value = softmax_loss(u_juicy, v_apple, V_size)
negative_sampling_loss_value = negative_sampling_loss(u_juicy, v_apple, K=5, V_size=V_size)

print(f"Softmax Loss: {softmax_loss_value}")
print(f"Negative Sampling Loss: {negative_sampling_loss_value}")

Softmax Loss: -0.0
Negative Sampling Loss: 5.132333592154672


Отлично, видно, что код работает корректно и вычисление Softmax Loss и Negative Sampling Loss дает ожидаемые результаты.

Небольшие пояснения по полученным значениям:

1. **Softmax Loss**: Значение -0.0 говорит о том, что вероятность `v_apple` при данном `u_juicy` близка к 1. Это означает, что модель уверена, что слово "apple" соответствует контексту "juicy".

2. **Negative Sampling Loss**: Значение 5.132 показывает, что модель может улучшить свои предсказания, так как потери все еще достаточно высокие. Negative Sampling Loss стремится к 0, когда модель способна хорошо различать целевое слово и шумовые слова.

Таким образом, мы успешно протестировали реализацию функций вычисления Softmax Loss и Negative Sampling Loss на простом примере. Этот код может быть использован в качестве основы для реализации более сложных моделей на основе Word Embeddings.

Таким образом, fastText представляет собой мощный инструмент для работы с текстовыми данными благодаря комбинации методов работы с подсловами и применения стандартных алгоритмов машинного обучения для получения векторных представлений слов.



**Doc2Vec**
Doc2Vec является методом для извлечения векторных представлений документов, основанным на архитектуре нейронных сетей, а именно на расширении модели Word2Vec для учета положения документов в пространстве.


### Введение в Doc2Vec

Doc2Vec, или Paragraph Vector, был представлен Квоком Ле и Томасом Миколовым в 2014 году. Этот метод является расширением идеи Word2Vec и позволяет получать векторные представления не только для отдельных слов, но и для целых документов, сохраняя их семантическую информацию.

#### 1. Архитектура Doc2Vec

Doc2Vec имеет две основные архитектуры: Distributed Memory (DM) и Distributed Bag of Words (DBOW). Давайте рассмотрим каждую из них подробнее.

#### 1.1 Distributed Memory (DM)

Архитектура Distributed Memory Doc2Vec включает в себя два основных вектора для каждого документа: один для представления самого документа в контексте других слов в документе (PV-DM), и один для учета общего контекста документа (DM).

##### PV-DM модель:

- В PV-DM модели для каждого документа $ d $ генерируется уникальный вектор $ \mathbf{d} $.
- Похожая на CBOW структура, где целевым словом является слово внутри документа, а контекстом — остальные слова из того же документа.
- Формула обновления вектора документа $ \mathbf{d} $:

$$ \mathbf{d}_t = \mathbf{d}_{t-1} + \eta \left( \sum_{w \in d} \frac{\partial \log P(w_t | \mathbf{d}_{t-1}, \mathbf{w}_{t-m}, ..., \mathbf{w}_{t+m})}{\partial \mathbf{d}_{t-1}} \right) $$

где $ \mathbf{d}_t $ — вектор документа на шаге $ t $, $ \mathbf{w}_{t-m}, ..., \mathbf{w}_{t+m} $ — контекст слова $ w_t $, $ \eta $ — скорость обучения.

##### DM модель:

- В DM модели вектор документа $ \mathbf{d} $ учитывает весь контекст документа.
- Документ рассматривается как контекст для генерации слов в рамках документа.
- Формула обновления вектора документа $ \mathbf{d} $:

$$ \mathbf{d}_t = \mathbf{d}_{t-1} + \eta \left( \frac{\partial \log P(w_t | \mathbf{d}_{t-1}, \mathbf{w}_{t-m}, ..., \mathbf{w}_{t+m})}{\partial \mathbf{d}_{t-1}} \right) $$

где $ \mathbf{d}_t $ — вектор документа на шаге $ t $, $ \mathbf{w}_{t-m}, ..., \mathbf{w}_{t+m} $ — контекст слова $ w_t $, $ \eta $ — скорость обучения.

#### 1.2 Distributed Bag of Words (DBOW)

В архитектуре DBOW нет внутреннего контекста документа, а только сам документ. DBOW представляет каждый документ одним вектором и предсказывает слова в нем независимо.

- Вектор документа $ \mathbf{d} $ в DBOW модели не учитывает контекст внутри документа.
- Формула обновления вектора документа $ \mathbf{d} $:

$$ \mathbf{d}_t = \mathbf{d}_{t-1} + \eta \left( \frac{\partial \log P(w_t | \mathbf{d}_{t-1})}{\partial \mathbf{d}_{t-1}} \right) $$

где $ \mathbf{d}_t $ — вектор документа на шаге $ t $, $ \eta $ — скорость обучения.

### 2. Обучение Doc2Vec

Обучение Doc2Vec требует достаточно больших объемов данных для получения хороших векторных представлений документов. Процесс обучения включает итеративное обновление векторов документов и слов в соответствии с выбранной архитектурой (DM или DBOW).

### 3. Оценка качества Doc2Vec

Оценка качества векторных представлений Doc2Vec может проводиться с помощью различных метрик, таких как:

- **Косинусное расстояние**: для оценки близости между векторами документов.
- **Классификационные задачи**: например, использование векторов для классификации текстовых документов.
- **Анализ кластеров**: оценка, как документы сгруппированы на основе их векторных представлений.

### Пример использования Doc2Vec

Посмотрим на простой пример использования Doc2Vec для анализа текстовых документов с помощью библиотеки `gensim`.



In [3]:
from gensim.models.doc2vec import Doc2Vec, TaggedDocument
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

# Пример данных (список текстовых документов)
documents = ["This is the first document", "Second document", "And another document"]

# Токенизация и создание объектов TaggedDocument
tagged_data = [TaggedDocument(words=word_tokenize(doc.lower()), tags=[str(i)]) for i, doc in enumerate(documents)]

# Обучение модели Doc2Vec (например, используя Distributed Memory модель)
model = Doc2Vec(tagged_data, vector_size=20, window=2, min_count=1, workers=4, epochs=100)

# Получение вектора для определенного документа
vector = model.infer_vector(word_tokenize("New document to infer"))

print("Vector representation of 'New document to infer':")
print(vector)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


Vector representation of 'New document to infer':
[ 0.02400572  0.00789973  0.00993905  0.02137663  0.02322021 -0.00634541
 -0.00072425  0.00107724  0.01944949 -0.01695026 -0.00344224 -0.01058895
  0.01041064  0.00196838 -0.01703393  0.00240276  0.01640063  0.01164591
 -0.00384032 -0.00034068]


Векторное представление документа "New document to infer", полученное с использованием модели Doc2Vec, представлено в виде массива чисел:

\[ [ 0.02400572,  0.00789973,  0.00993905,  0.02137663,  0.02322021, -0.00634541, -0.00072425,  0.00107724,  0.01944949, -0.01695026, -0.00344224, -0.01058895,  0.01041064,  0.00196838, -0.01703393,  0.00240276,  0.01640063,  0.01164591, -0.00384032, -0.00034068 ] \]

Этот вектор содержит числовые значения, которые представляют семантическое содержание документа "New document to infer" в контексте всего корпуса текстов, на котором была обучена модель Doc2Vec. Каждая компонента этого вектора представляет собой числовое значение, отражающее семантические аспекты этого документа.

Эти векторы могут использоваться для различных задач анализа текста, таких как поиск похожих документов, классификация текстов, кластеризация документов и другие. Важно отметить, что точное значение каждой компоненты вектора зависит от конкретной модели Doc2Vec и параметров её обучения.

В данном разделе мы рассмотрели основные аспекты метода Doc2Vec, включая его архитектуры (DM и DBOW), процесс обучения, оценку качества и примеры использования. Doc2Vec является мощным инструментом для извлечения семантических векторных представлений документов, что позволяет эффективно работать с текстовыми данными в различных задачах анализа и обработки текста.

**1.3 Создание векторов тем**

Темы могут быть представлены как комбинации векторов слов. Для этого часто используются методы Latent Dirichlet Allocation (LDA) и Latent Semantic Analysis (LSA):
1. **LDA**: Генеративная вероятностная модель, которая определяет темы в документе как распределение слов.
2. **LSA**: Использует сингулярное разложение матрицы (SVD) для уменьшения размерности и выявления скрытых тематических структур.



#### Latent Semantic Analysis (LSA)

Latent Semantic Analysis (LSA) — это метод анализа текстов, который используется для выявления семантической структуры документов путем анализа статистических связей между терминами и документами. Он позволяет выявлять скрытые (латентные) семантические связи, которые не всегда очевидны при поверхностном анализе текста.

#### Основные принципы работы LSA

1. **Представление текста в виде матрицы термин-документ**

   LSA начинается с построения матрицы термин-документ, где каждый элемент $ A_{ij} $ представляет степень важности термина $ i $ в документе $ j $. Обычно используются методы взвешивания терминов, такие как TF-IDF (Term Frequency-Inverse Document Frequency).

   Пусть у нас есть коллекция документов $ D $, состоящая из $ N $ документов, и словарь терминов $ T $, состоящий из $ M $ уникальных терминов. Матрица термин-документ будет иметь размерность $ M \times N $.

2. **Сингулярное разложение (SVD)**

   Основная идея LSA заключается в применении сингулярного разложения матрицы $ A $, чтобы снизить размерность и извлечь латентные семантические структуры.

   Пусть $ A $ — это матрица $ M \times N $ термин-документ, где $ M $ — количество терминов, $ N $ — количество документов. Сингулярное разложение $ A $ выглядит следующим образом:

   $$ A = U \Sigma V^T $$

   где:
   - $ U $ — ортогональная матрица размерности $ M \times M $,
   - $ \Sigma $ — диагональная матрица размерности $ M \times N $ с сингулярными значениями,
   - $ V $ — ортогональная матрица размерности $ N \times N $.

3. **Выбор размерности сокращения**

   Одним из ключевых шагов в LSA является выбор размерности $ k $ для сокращения. Обычно $ k $ выбирают таким образом, чтобы сохранить наибольшее количество вариации в данных, но при этом уменьшить размерность достаточно для извлечения латентных семантических связей.

#### Пример работы LSA

Рассмотрим пример коллекции документов:

- Документ 1: "Кошка и собака — домашние животные."
- Документ 2: "Кошка часто спит."
- Документ 3: "Собаки любят гулять."

Шаги LSA для этого примера:

1. **Построение матрицы термин-документ**

   После предобработки текстов и применения метода взвешивания (например, TF-IDF) получаем матрицу $ A $:

   $$
   A = \begin{bmatrix}
   1 & 1 & 1 \\
   1 & 0 & 0 \\
   0 & 1 & 1 \\
   \end{bmatrix}
   $$

   где строки соответствуют терминам (кошка, собака, домашние, животные, часто, спит, любят, гулять), а столбцы — документам.

2. **Применение сингулярного разложения (SVD)**

   Выполняем SVD для матрицы $ A $:

   $$
   A = U \Sigma V^T
   $$

   где $ U $, $ \Sigma $ и $ V $ — ортогональные матрицы, а $ \Sigma $ содержит сингулярные значения.

3. **Выбор размерности $ k $**

   После выполнения SVD выбираем размерность $ k $ для сокращения.

4. **Интерпретация результатов**

   После снижения размерности до $ k $ можно проанализировать полученные тематические компоненты и сравнить документы по их семантическому содержанию.

Таким, образом LSA является мощным методом для анализа семантической структуры текстовых данных. Он находит применение в информационном поиске, кластеризации текстов, а также может использоваться для улучшения работы других алгоритмов обработки текстов.


#### Пример работы LSA

Рассмотрим следующий пример с тремя документами и тремя уникальными терминами:

- Документ 1: "Кошка собака"
- Документ 2: "Кошка спит"
- Документ 3: "Собака гуляет"

**Шаг 1: Построение матрицы термин-документ**

Пусть термины "кошка", "собака" и "спит" будут обозначены как $ t_1 $, $ t_2 $ и $ t_3 $ соответственно. Матрица термин-документ $ A $ будет иметь размер $ 3 \times 3 $:

$$
A = \begin{pmatrix}
1 & 1 & 0 \\
1 & 0 & 1 \\
0 & 1 & 0 \\
\end{pmatrix}
$$

Здесь строки соответствуют терминам, а столбцы — документам.

**Шаг 2: Применение сингулярного разложения (SVD)**

Проводим сингулярное разложение для матрицы $ A $:

$$
A = U \Sigma V^T
$$

Где:
- $ U $ — ортогональная матрица размером $ 3 \times 3 $,
- $ \Sigma $ — диагональная матрица размером $ 3 \times 3 $ с сингулярными значениями на диагонали,
- $ V $ — ортогональная матрица размером $ 3 \times 3 $.

В результате разложения получаем:

$$
U = \begin{pmatrix}
-0.707 & 0.000 & 0.707 \\
-0.707 & 0.000 & -0.707 \\
0.000 & 1.000 & 0.000 \\
\end{pmatrix}
$$

$$
\Sigma = \begin{pmatrix}
1.414 & 0.000 & 0.000 \\
0.000 & 1.000 & 0.000 \\
0.000 & 0.000 & 0.000 \\
\end{pmatrix}
$$

$$
V^T = \begin{pmatrix}
-0.707 & -0.707 & 0.000 \\
0.707 & -0.707 & 0.000 \\
0.000 & 0.000 & 1.000 \\
\end{pmatrix}
$$

**Шаг 3: Выбор размерности $ k $**

Для упрощения и иллюстрации выбираем $ k = 2 $, сохраняя два наибольших сингулярных значения. Обрезаем $ \Sigma $ до размерности $ 2 \times 2 $, а $ U $ и $ V $ до $ 3 \times 2 $ и $ 2 \times 3 $ соответственно:

$$
U_k = \begin{pmatrix}
-0.707 & 0.000 \\
-0.707 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
$$

$$
\Sigma_k = \begin{pmatrix}
1.414 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
$$

$$
V^T_k = \begin{pmatrix}
-0.707 & -0.707 & 0.000 \\
0.707 & -0.707 & 0.000 \\
\end{pmatrix}
$$

**Шаг 4: Интерпретация результатов**

Новая матрица термин-документ после снижения размерности будет:

$$
A_k = U_k \Sigma_k V^T_k
$$

$$
A_k = \begin{pmatrix}
-0.707 & 0.000 \\
-0.707 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
\begin{pmatrix}
1.414 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
\begin{pmatrix}
-0.707 & -0.707 & 0.000 \\
0.707 & -0.707 & 0.000 \\
\end{pmatrix}
$$

Умножаем эти матрицы:

$$
A_k = \begin{pmatrix}
-0.707 \cdot 1.414 + 0.000 \cdot 0.707 & -0.707 \cdot 1.414 + 0.000 \cdot -0.707 & 0.000 \\
-0.707 \cdot 1.414 + 0.000 \cdot 0.707 & -0.707 \cdot 1.414 + 0.000 \cdot -0.707 & 0.000 \\
0.000 \cdot 1.414 + 1.000 \cdot 0.707 & 0.000 \cdot -0.707 + 1.000 \cdot -0.707 & 0.000 \\
\end{pmatrix}
$$

$$
A_k = \begin{pmatrix}
-1.000 & -1.000 & 0.000 \\
-1.000 & -1.000 & 0.000 \\
0.707 & -0.707 & 0.000 \\
\end{pmatrix}
$$

Эта матрица $ A_k $ представляет документы в новом пространстве меньшей размерности, где латентные семантические связи между терминами и документами более явные. Например, первый и второй документы теперь ближе друг к другу, что отражает их семантическое сходство ("кошка" упоминается в обоих).

Таким образом, LSA позволяет эффективно выявлять скрытые семантические связи в текстовых данных. Применение SVD и снижение размерности помогают обнаруживать латентные структуры, улучшая качество анализа текстов.


In [ ]:
!pip install stop-words
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

# Пример коллекции текстов (новостные заголовки)
documents = [
    "Турция выиграла матч в чемпионате мира",
    "Россия проиграла в полуфинале чемпионата Европы",
    "США победили в Олимпийских играх",
    "Франция вышла в финал Лиги наций",
    "Италия выиграла золото в чемпионате мира"
]

# Получение списка стоп-слов на русском языке
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
stop_words = list(ENGLISH_STOP_WORDS) + ['и', 'в', 'на', 'не', 'с', 'от', 'для', 'с', 'по', 'из', 'что', 'так', 'как', 'это', 'этот']

# Инициализируем TF-IDF векторизатор с указанием стоп-слов
vectorizer = TfidfVectorizer(stop_words=stop_words)
X = vectorizer.fit_transform(documents)

# Применяем сингулярное разложение (SVD) для уменьшения размерности до 2
lsa = TruncatedSVD(n_components=2)
X_reduced = lsa.fit_transform(X)

# Выводим результаты
print("Original TF-IDF matrix shape:", X.shape)
print("Reduced matrix shape:", X_reduced.shape)
print("Singular values:", lsa.singular_values_)
print()

# Выводим топ слов для каждой компоненты
terms = vectorizer.get_feature_names_out()
for i, component in enumerate(lsa.components_):
    top_terms = [terms[ind] for ind in component.argsort()[:-6:-1]]
    print(f"Top terms for component {i+1}: {', '.join(top_terms)}")
    print()

# Выводим преобразованные документы (документы в новом пространстве)
print("Transformed documents (in reduced space):")
for i, text in enumerate(documents):
    print(f"Document {i+1}: {X_reduced[i]}")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Original TF-IDF matrix shape: (5, 21)
Reduced matrix shape: (5, 2)
Singular values: [1.2223023 1.       ]

Top terms for component 1: выиграла, мира, чемпионате, золото, италия

Top terms for component 2: россия, проиграла, полуфинале, чемпионата, европы

Transformed documents (in reduced space):
Document 1: [8.64298248e-01 8.20954029e-17]
Document 2: [2.5419027e-16 9.6969697e-01]
Document 3: [-1.31004070e-16 -2.12121212e-01]
Document 4: [6.43022749e-17 1.21212121e-01]
Document 5: [ 8.64298248e-01 -3.20524508e-16]


### Latent Dirichlet Allocation (LDA)

Latent Dirichlet Allocation (LDA) — это вероятностная модель, используемая для обнаружения скрытых тем в коллекции текстовых документов. LDA представляет документы как смеси различных тем, где каждая тема характеризуется распределением терминов. Этот метод широко применяется в задачах тематического моделирования, текстовой классификации и информационного поиска.

#### Основные понятия и предположения LDA

1. **Темы**: Каждая тема — это распределение вероятностей на словарных терминах.
2. **Документы**: Каждый документ рассматривается как смесь тем, и каждая тема в документе имеет свою долю.
3. **Слова**: Каждое слово в документе связано с одной из тем.

#### Графическая модель LDA

LDA можно представить в виде графической модели, где:
- $\alpha$ — гиперпараметр, управляющий распределением тем в документах.
- $\beta$ — гиперпараметр, управляющий распределением слов в темах.
- $\theta_d$ — распределение тем для документа $d$.
- $\phi_k$ — распределение слов для темы $k$.
- $z_{d,n}$ — тема для $n$-го слова в документе $d$.
- $w_{d,n}$ — $n$-ое слово в документе $d$.

Графическая модель для одного документа выглядит следующим образом:

$$
\begin{array}{ccccccc}
\alpha & & \beta & & & & \\
& \theta_d & & \phi_k & & & \\
& \downarrow & & \downarrow & & & \\
\text{(Dir)} & & \text{(Dir)} & & & & \\
& & z_{d,n} & \rightarrow & w_{d,n} & & \\
\end{array}
$$

#### Основные шаги LDA

1. **Генерация распределений тем для каждого документа**:
   - Для каждого документа $d$ выбирается распределение тем $\theta_d \sim \text{Dir}(\alpha)$.

2. **Генерация распределений слов для каждой темы**:
   - Для каждой темы $k$ выбирается распределение слов $\phi_k \sim \text{Dir}(\beta)$.

3. **Генерация слов для документов**:
   - Для каждого слова $n$ в документе $d$:
     - Выбирается тема $z_{d,n} \sim \text{Multinomial}(\theta_d)$.
     - Выбирается слово $w_{d,n}$ согласно распределению слов для темы $z_{d,n}$, то есть $w_{d,n} \sim \text{Multinomial}(\phi_{z_{d,n}})$.

#### Формулы LDA

1. **Вероятность темы $z_{d,n}$ для слова $w_{d,n}$**:
   $$
   P(z_{d,n} = k | w_{d,n} = v, \theta_d, \phi_k) \propto P(w_{d,n} = v | z_{d,n} = k, \phi_k) P(z_{d,n} = k | \theta_d)
   $$

2. **Совместная вероятность модели**:
   $$
   P(w, z | \alpha, \beta) = \prod_{d=1}^D \left[ P(\theta_d | \alpha) \prod_{n=1}^{N_d} P(z_{d,n} | \theta_d) P(w_{d,n} | z_{d,n}, \phi) \right]
   $$

3. **Маргинальная вероятность документа $d$**:
   $$
   P(w_d | \alpha, \beta) = \int P(\theta_d | \alpha) \left( \prod_{n=1}^{N_d} \sum_{z_{d,n}} P(z_{d,n} | \theta_d) P(w_{d,n} | z_{d,n}, \beta) \right) d\theta_d
   $$

#### Пример работы LDA

Рассмотрим пример с тремя документами и двумя темами.

1. **Документы**:
   - Документ 1: "Кошка играет с мячом"
   - Документ 2: "Собака любит бегать"
   - Документ 3: "Кошка и собака играют вместе"

2. **Предположим две темы**:
   - Тема 1: Домашние животные
   - Тема 2: Игры и активность

3. **Процесс генерации**:

   - **Шаг 1**: Определим распределение тем для каждого документа ($\theta$):
     - Документ 1: $\theta_1 = [0.6, 0.4]$ (60% о домашних животных, 40% об играх)
     - Документ 2: $\theta_2 = [0.3, 0.7]$ (30% о домашних животных, 70% об играх)
     - Документ 3: $\theta_3 = [0.5, 0.5]$ (50% о домашних животных, 50% об играх)

   - **Шаг 2**: Определим распределение слов для каждой темы ($\phi$):
     - Тема 1: $\phi_1 = [0.3, 0.3, 0.2, 0.2]$ (слова: "кошка", "собака", "играет", "бегает")
     - Тема 2: $\phi_2 = [0.1, 0.1, 0.4, 0.4]$ (слова: "кошка", "собака", "играет", "бегает")

   - **Шаг 3**: Генерация слов для каждого документа:
     - Для документа 1:
       - Выбираем тему для первого слова, например, Тема 1.
       - Выбираем слово из Тема 1, например, "кошка".
       - Повторяем процесс для всех слов в документе.

#### Оценка параметров в LDA

Для оценки параметров $\theta$ и $\phi$ используются различные методы, включая вариационный байесовский вывод и метод Гиббсовой выборки.

1. **Гиббсовая выборка**:
   - Итеративный метод для аппроксимации распределений $\theta$ и $\phi$.
   - Обновляются условные распределения $P(z_{d,n} | w, z_{-d,n})$, где $z_{-d,n}$ — это все переменные $z$ кроме $z_{d,n}$.

2. **Вариационный байесовский вывод**:
   - Метод оптимизации для аппроксимации распределений $\theta$ и $\phi$.
   - Заменяет сложные распределения на более простые аппроксимации и минимизирует расстояние между ними.

#### Применение LDA

LDA широко применяется в различных областях, таких как:

1. **Тематическое моделирование**:
   - Обнаружение тем в больших коллекциях текстов, таких как научные статьи, новости, блоги.

2. **Текстовая классификация**:
   - Автоматическая классификация документов по темам.

3. **Информационный поиск**:
   - Улучшение релевантности поиска путем учета скрытых тем в документах.

4. **Рекомендательные системы**:
   - Рекомендация контента пользователям на основе их интересов, определенных через темы.




#### Конкретный пример работы LDA

Рассмотрим пример с тремя документами и двумя темами.

##### Документы:
- Документ 1: "Кошка играет с мячом"
- Документ 2: "Собака любит бегать"
- Документ 3: "Кошка и собака играют вместе"

##### Шаг 1: Построение матрицы термин-документ

Предположим, у нас есть следующие термины: "кошка", "играет", "мяч", "собака", "любит", "бегать", "вместе".

Матрица термин-документ:

$$
\begin{array}{c|ccc}
 & \text{Документ 1} & \text{Документ 2} & \text{Документ 3} \\
\hline
\text{кошка} & 1 & 0 & 1 \\
\text{играет} & 1 & 0 & 1 \\
\text{мяч} & 1 & 0 & 0 \\
\text{собака} & 0 & 1 & 1 \\
\text{любит} & 0 & 1 & 0 \\
\text{бегать} & 0 & 1 & 0 \\
\text{вместе} & 0 & 0 & 1 \\
\end{array}
$$

##### Шаг 2: Генерация распределений тем для каждого документа

Для упрощения примера, предположим следующие значения параметров:

- Гиперпараметр $\alpha$ для распределения тем: $\alpha = [0.5, 0.5]$
- Гиперпараметр $\beta$ для распределения слов: $\beta = [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]$

Распределение тем для каждого документа ($\theta$) может быть следующим:

- Документ 1: $\theta_1 = [0.6, 0.4]$ (60% Тема 1, 40% Тема 2)
- Документ 2: $\theta_2 = [0.3, 0.7]$ (30% Тема 1, 70% Тема 2)
- Документ 3: $\theta_3 = [0.5, 0.5]$ (50% Тема 1, 50% Тема 2)

##### Шаг 3: Генерация распределений слов для каждой темы

Распределение слов для каждой темы ($\phi$) может быть следующим:

- Тема 1: $\phi_1 = [0.3, 0.3, 0.2, 0.1, 0.05, 0.025, 0.025]$
- Тема 2: $\phi_2 = [0.1, 0.1, 0.1, 0.4, 0.1, 0.1, 0.1]$

##### Шаг 4: Генерация слов для документов

Для каждого слова $n$ в документе $d$:

- Выбираем тему $z_{d,n}$ из распределения тем $\theta_d$.
- Выбираем слово $w_{d,n}$ из распределения слов $\phi_{z_{d,n}}$.

Рассмотрим пример для Документа 1: "Кошка играет с мячом".

1. Первое слово "кошка":
   - Выбираем тему: $z_{1,1} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 1.
   - Выбираем слово из Тема 1: $w_{1,1} \sim \text{Multinomial}(\phi_1)$. Пусть выбрано "кошка".

2. Второе слово "играет":
   - Выбираем тему: $z_{1,2} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 1.
   - Выбираем слово из Тема 1: $w_{1,2} \sim \text{Multinomial}(\phi_1)$. Пусть выбрано "играет".

3. Третье слово "с":
   - Выбираем тему: $z_{1,3} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 2.
   - Выбираем слово из Тема 2: $w_{1,3} \sim \text{Multinomial}(\phi_2)$. Пусть выбрано "с".

4. Четвертое слово "мячом":
   - Выбираем тему: $z_{1,4} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 1.
   - Выбираем слово из Тема 1: $w_{1,4} \sim \text{Multinomial}(\phi_1)$. Пусть выбрано "мячом".

#### Оценка параметров в LDA

Для оценки параметров $\theta$ и $\phi$ используются различные методы, включая вариационный байесовский вывод и метод Гиббсовой выборки.

1. **Гиббсовая выборка**:
   - Итеративный метод для аппроксимации распределений $\theta$ и $\phi$.
   - Обновляются условные распределения $P(z_{d,n} | w, z_{-d,n})$, где $z_{-d,n}$ — это все переменные $z$ кроме $z_{d,n}$.

2. **Вариационный байесовский вывод**:
   - Метод оптимизации для аппроксимации распределений $\theta$ и $\phi$.
   - Заменяет сложные распределения на более простые аппроксимации и минимизирует расстояние между ними.

Таким образом, Latent Dirichlet Allocation (LDA) — мощный инструмент для анализа текстов, позволяющий выявлять скрытые семантические структуры и улучшать обработку текстовых данных. Вероятностный подход и применение байесовских методов делают LDA гибким и эффективным для различных задач обработки естественного языка.


In [ ]:
!pip install pyLDAvis

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 14.8 MB/s eta 0:00:00


In [ ]:
import gensim
from gensim import corpora
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import nltk

# Скачиваем ресурсы NLTK
nltk.download('punkt')
nltk.download('stopwords')

# Пример документов
documents = [
    "Кошка играет с мячом",
    "Собака любит бегать",
    "Кошка и собака играют вместе"
]

# Предобработка текстов
stop_words = set(stopwords.words('russian'))
texts = [
    [word for word in word_tokenize(document.lower()) if word.isalnum() and word not in stop_words]
    for document in documents
]

# Создание словаря и корпуса
dictionary = corpora.Dictionary(texts)
corpus = [dictionary.doc2bow(text) for text in texts]

# Параметры LDA
num_topics = 2  # Количество тем
passes = 10     # Количество проходов для улучшения модели

# Обучение модели LDA
lda_model = gensim.models.LdaModel(corpus, num_topics=num_topics, id2word=dictionary, passes=passes)

# Вывод тем
for idx, topic in lda_model.print_topics(-1):
    print(f"Тема {idx+1}: {topic}")

# Визуализация тем с pyLDAvis
import pyLDAvis
import pyLDAvis.gensim_models as gensimvis

# Визуализация LDA модели
lda_display = gensimvis.prepare(lda_model, corpus, dictionary, sort_topics=False)

# Сохранение визуализации в файл HTML
pyLDAvis.save_html(lda_display, 'lda_vis.html')

# Сообщение для пользователя
print("Визуализация сохранена в файл lda_vis.html. Откройте его в браузере для просмотра.")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Тема 1: 0.219*"кошка" + 0.209*"мячом" + 0.208*"играет" + 0.074*"собака" + 0.073*"играют" + 0.073*"вместе" + 0.073*"любит" + 0.072*"бегать"
Тема 2: 0.228*"собака" + 0.137*"бегать" + 0.137*"любит" + 0.137*"вместе" + 0.136*"играют" + 0.132*"кошка" + 0.047*"играет" + 0.047*"мячом"
Визуализация сохранена в файл lda_vis.html. Откройте его в браузере для просмотра.


Формула представления результатов модели LDA, представляет собой веса ключевых слов для каждой темы. Каждая строка соответствует одной теме и содержит ключевые слова в порядке убывания их важности для данной темы, а также их веса.

Например, для первой темы:
- "кошка" имеет вес 0.219
- "мячом" имеет вес 0.209
- "играет" имеет вес 0.208
- и так далее...

Эти веса указывают на значимость каждого слова в контексте данной темы. Чем выше вес, тем более важным является это слово для определения темы. Если у вас есть еще вопросы или нужна дополнительная информация, не стесняйтесь спрашивать!


##### Пример 2: Тематическое моделирование новостных статей

Представим, что у нас есть коллекция новостных статей, и мы хотим выделить основные темы, обсуждаемые в этих статьях. Мы можем использовать Latent Dirichlet Allocation (LDA) для моделирования скрытых тем.

Пример текста новостной статьи:

```
Новость 1: "Вчера в Париже состоялась встреча лидеров Г7. Основные темы обсуждения были связаны с изменением климата и международной безопасностью."
Новость 2: "В Кремле состоялась встреча президентов России и Франции. Главные темы обсуждения включали ситуацию на Украине и торговые отношения."
Новость 3: "На Silicon Valley прошла конференция по искусственному интеллекту. Основные выступления были посвящены автономным транспортным средствам и глубокому обучению."
```

После применения LDA мы можем выделить следующие темы:
- Тема 1: Международные отношения и безопасность
- Тема 2: Искусственный интеллект и технологии
- Тема 3: Климатические изменения и экология

Это помогает быстро оценить, о чём идёт речь в новостной коллекции, и определить, какие темы в данный момент находятся в центре внимания общественности и СМИ.


Для реализации тематического моделирования новостных статей с использованием Latent Dirichlet Allocation (LDA) в Python нам понадобятся следующие шаги:

### Шаги реализации:

#### 1. Подготовка данных

Сначала мы определим коллекцию новостных статей в виде текстов. Каждая статья будет представлена строкой текста.

```python
# Пример текстов новостных статей
news_articles = [
    "Вчера в Париже состоялась встреча лидеров Г7. Основные темы обсуждения были связаны с изменением климата и международной безопасностью.",
    "В Кремле состоялась встреча президентов России и Франции. Главные темы обсуждения включали ситуацию на Украине и торговые отношения.",
    "На Silicon Valley прошла конференция по искусственному интеллекту. Основные выступления были посвящены автономным транспортным средствам и глубокому обучению."
]
```

#### 2. Токенизация и предобработка текста

Тексты новостных статей необходимо токенизировать (разделить на отдельные слова или токены) и провести базовую предобработку, такую как удаление стоп-слов, приведение к нижнему регистру и т.д. Для этого можно использовать библиотеку `nltk`.

```python
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
nltk.download('punkt')
nltk.download('stopwords')

# Удаление стоп-слов и лемматизация
stop_words = set(stopwords.words('russian'))
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    tokens = word_tokenize(text.lower())
    tokens = [token for token in tokens if token.isalpha()]  # удаление пунктуации
    tokens = [token for token in tokens if token not in stop_words]
    tokens = [lemmatizer.lemmatize(token) for token in tokens]
    return tokens

processed_articles = [preprocess_text(article) for article in news_articles]
```

#### 3. Создание словаря и корпуса

Для работы с LDA необходимо создать словарь всех уникальных токенов и представить каждую статью в виде мешка слов (bag of words).

```python
from gensim.corpora import Dictionary
from gensim.models import LdaModel

# Создание словаря
dictionary = Dictionary(processed_articles)

# Создание корпуса (представление каждой статьи в виде мешка слов)
corpus = [dictionary.doc2bow(article) for article in processed_articles]
```

#### 4. Обучение модели LDA

Теперь мы можем обучить модель LDA на нашем корпусе. Важно указать количество тем (k), которые мы хотим выделить.

```python
# Обучение модели LDA
lda_model = LdaModel(corpus, num_topics=3, id2word=dictionary, passes=10)
```

#### 5. Интерпретация результатов

После обучения модели мы можем вывести темы и их наиболее вероятные слова.

```python
# Вывод тем и их ключевых слов
topics = lda_model.print_topics(num_words=5)
for topic in topics:
    print(topic)
```

### Полный код

Ниже представлен полный код для реализации тематического моделирования новостных статей с использованием LDA:



In [6]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from gensim.corpora import Dictionary
from gensim.models import LdaModel
import nltk
nltk.download('wordnet')

# Пример текстов новостных статей
news_articles = [
    "Вчера в Париже состоялась встреча лидеров Г7. Основные темы обсуждения были связаны с изменением климата и международной безопасностью.",
    "В Кремле состоялась встреча президентов России и Франции. Главные темы обсуждения включали ситуацию на Украине и торговые отношения.",
    "На Silicon Valley прошла конференция по искусственному интеллекту. Основные выступления были посвящены автономным транспортным средствам и глубокому обучению."
]

# Удаление стоп-слов и лемматизация
nltk.download('punkt')
nltk.download('stopwords')
stop_words = set(stopwords.words('russian'))
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    tokens = word_tokenize(text.lower())
    tokens = [token for token in tokens if token.isalpha()]  # удаление пунктуации
    tokens = [token for token in tokens if token not in stop_words]
    tokens = [lemmatizer.lemmatize(token) for token in tokens]
    return tokens

processed_articles = [preprocess_text(article) for article in news_articles]

# Создание словаря
dictionary = Dictionary(processed_articles)

# Создание корпуса (представление каждой статьи в виде мешка слов)
corpus = [dictionary.doc2bow(article) for article in processed_articles]

# Обучение модели LDA
lda_model = LdaModel(corpus, num_topics=3, id2word=dictionary, passes=10)

# Вывод тем и их ключевых слов
topics = lda_model.print_topics(num_words=5)
for topic in topics:
    print(topic)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


(0, '0.060*"основные" + 0.034*"вчера" + 0.034*"изменением" + 0.034*"лидеров" + 0.034*"связаны"')
(1, '0.051*"состоялась" + 0.051*"темы" + 0.051*"обсуждения" + 0.051*"встреча" + 0.051*"включали"')
(2, '0.028*"встреча" + 0.028*"обсуждения" + 0.028*"торговые" + 0.028*"темы" + 0.028*"россии"')


Эти строки представляют результаты тем, выделенных моделью LDA (Latent Dirichlet Allocation), с их ключевыми словами и их весами. Давайте проинтерпретируем каждую тему на основе представленных ключевых слов:

### Тема 0:
- Ключевые слова: "основные", "вчера", "изменением", "лидеров", "связаны"
- Интерпретация: Вероятно, эта тема связана с обсуждением ключевых тем или аспектов, которые являются основными или важными. Упоминание "вчера" может указывать на временной контекст или на то, что обсуждение происходило недавно. "Изменением" может относиться к изменениям, происходящим в контексте темы.

### Тема 1:
- Ключевые слова: "состоялась", "темы", "обсуждения", "встреча", "включали"
- Интерпретация: Эта тема, вероятно, относится к событиям, таким как встречи, обсуждения и ключевые темы, которые были включены в обсуждения. Слова "состоялась" и "встреча" указывают на проведение событий или встреч, в ходе которых происходили обсуждения.

### Тема 2:
- Ключевые слова: "встреча", "обсуждения", "торговые", "темы", "россии"
- Интерпретация: Эта тема, возможно, связана с обсуждением встреч, торговых отношений и ключевых тем, касающихся России. Упоминание "торговые" указывает на экономические аспекты, а "россии" — на страновые аспекты или упоминания.

### Общие замечания:
- Вес каждого ключевого слова в теме отражает его значимость или частоту в данной теме по отношению к другим словам в этой теме.
- Результаты LDA могут варьироваться в зависимости от параметров модели и предобработки текста.
- Для лучшего понимания тем можно рассмотреть больше ключевых слов или использовать более глубокий анализ результатов моделирования.

Эти темы помогают кратко описать основные аспекты, которые обсуждаются в коллекции новостных статей, и могут быть полезны для анализа содержания и автоматизированного каталогизирования новостей.

#### 2. Семантический поиск на основе сходства векторов тем



## 2.1 Понятие семантического сходства

Семантическое сходство измеряет близость значений слов, предложений или документов. Для этого используются различные метрики, которые оценивают расстояние или угол между их векторными представлениями.

### Метрики семантического сходства

#### Косинусное сходство

Косинусное сходство измеряет косинус угла между двумя векторами. Формула косинусного сходства между двумя векторами $ \mathbf{A} $ и $ \mathbf{B} $ выглядит следующим образом:

$$ \text{cosine_similarity}(\mathbf{A}, \mathbf{B}) = \frac{\mathbf{A} \cdot \mathbf{B}}{\|\mathbf{A}\| \|\mathbf{B}\|} = \frac{\sum_{i=1}^{n} A_i B_i}{\sqrt{\sum_{i=1}^{n} A_i^2} \sqrt{\sum_{i=1}^{n} B_i^2}} $$

где $ \mathbf{A} \cdot \mathbf{B} $ — скалярное произведение векторов, а $ \|\mathbf{A}\| $ и $ \|\mathbf{B}\| $ — их нормы (длины).

#### Евклидово расстояние

Евклидово расстояние измеряет прямую линию между двумя точками (векторами) в пространстве. Формула для двух векторов $ \mathbf{A} $ и $ \mathbf{B} $ выглядит так:

$$ \text{euclidean_distance}(\mathbf{A}, \mathbf{B}) = \sqrt{\sum_{i=1}^{n} (A_i - B_i)^2} $$

#### Расстояние Минковского

Расстояние Минковского обобщает Евклидово расстояние и Манхэттенское расстояние. Формула для векторов $ \mathbf{A} $ и $ \mathbf{B} $:

$$ \text{minkowski_distance}(\mathbf{A}, \mathbf{B}, p) = \left( \sum_{i=1}^{n} |A_i - B_i|^p \right)^{1/p} $$

где $ p $ — параметр, определяющий тип расстояния: $ p=1 $ — Манхэттенское расстояние, $ p=2 $ — Евклидово расстояние.

### Примеры

- Для векторов $ \mathbf{A} = [1, 2, 3] $ и $ \mathbf{B} = [4, 5, 6] $:
  - Косинусное сходство:

  $$
  \text{cosine_similarity}([1, 2, 3], [4, 5, 6]) = \frac{1 \cdot 4 + 2 \cdot 5 + 3 \cdot 6}{\sqrt{1^2 + 2^2 + 3^2} \sqrt{4^2 + 5^2 + 6^2}} = \frac{32}{\sqrt{14} \sqrt{77}} \approx 0.9746
  $$

  - Евклидово расстояние:

  $$
  \text{euclidean_distance}([1, 2, 3], [4, 5, 6]) = \sqrt{(1-4)^2 + (2-5)^2 + (3-6)^2} = \sqrt{9 + 9 + 9} = \sqrt{27} \approx 5.196
  $$

  - Расстояние Минковского при $ p=3 $:

  $$
  \text{minkowski_distance}([1, 2, 3], [4, 5, 6], 3) = \left( |1-4|^3 + |2-5|^3 + |3-6|^3 \right)^{1/3} = \left( 27 + 27 + 27 \right)^{1/3} = \left( 81 \right)^{1/3} \approx 4.326
  $$

## 2.2 Алгоритмы семантического поиска

### 1. Поиск на основе векторов слов

Для поиска на основе векторов слов используются модели эмбеддинга, такие как Word2Vec, GloVe или FastText. Эти модели преобразуют слова в векторы фиксированной размерности, в которых слова с похожими значениями располагаются ближе друг к другу в векторном пространстве.

#### Пример использования Word2Vec

Обученная модель Word2Vec может предоставить векторные представления для слов. Пусть у нас есть два слова: "king" и "queen". Мы можем получить их векторные представления и использовать косинусное сходство для оценки их семантического сходства:

$$
\text{cosine_similarity}(\mathbf{king}, \mathbf{queen}) = \frac{\mathbf{king} \cdot \mathbf{queen}}{\|\mathbf{king}\| \|\mathbf{queen}\|}
$$

### 2. Поиск на основе векторов предложений или документов

Для получения векторных представлений целых предложений или документов используются более сложные модели, такие как BERT, Sentence-BERT или Universal Sentence Encoder. Эти модели учитывают контекст слов и создают эмбеддинги, которые можно использовать для семантического поиска.

#### Пример использования BERT

Модель BERT может быть использована для создания эмбеддингов предложений. Пусть у нас есть два предложения: "The cat sits on the mat" и "The dog lies on the rug". Мы можем получить их векторные представления и использовать косинусное сходство для оценки их семантического сходства:

1. Преобразуем предложения в векторные представления с помощью модели BERT.
2. Вычислим косинусное сходство между полученными векторами.

### Примеры

#### Получение эмбеддингов с помощью Sentence-BERT

Используем модель Sentence-BERT для получения эмбеддингов:


In [ ]:
!pip install sentence_transformers
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

sentences = ["The cat sits on the mat", "The dog lies on the rug"]
embeddings = model.encode(sentences)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 227.1/227.1 kB 2.2 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-m

/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:11: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.73k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/314 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Теперь у нас есть векторные представления для обоих предложений. Вычислим косинусное сходство между ними:

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

similarity = cosine_similarity([embeddings[0]], [embeddings[1]])
print(similarity)

[[0.36165166]]


Этот код выведет значение косинусного сходства между двумя предложениями.

Таким боразом, семантический поиск на основе сходства векторов тем позволяет находить похожие документы, предложения или слова, оценивая их семантическое сходство. Использование различных метрик, таких как косинусное сходство, евклидово расстояние и расстояние Минковского, позволяет гибко подходить к задаче оценки сходства. Современные модели эмбеддинга, такие как BERT и Sentence-BERT, предоставляют мощные инструменты для получения качественных векторных представлений текста, что существенно улучшает результаты семантического поиска.

## 2.3 Применение в реальных системах

Семантический поиск на основе сходства векторов тем находит широкое применение в различных областях, таких как информационный поиск и рекомендательные системы. В этой лекции мы подробно рассмотрим, как семантический анализ используется для улучшения поиска и персонализации рекомендаций.

### 1. Информационный поиск

Информационный поиск включает нахождение релевантной информации в больших объемах данных. Поисковые системы, такие как Google или Bing, используют семантический анализ для улучшения качества поиска.

#### Принципы работы

1. **Предварительная обработка данных**: Текст документов и запросов пользователей предварительно обрабатывается (токенизация, лемматизация, удаление стоп-слов и т.д.).
2. **Преобразование в векторы**: Текст преобразуется в векторные представления с помощью моделей эмбеддинга (Word2Vec, GloVe, BERT и др.).
3. **Оценка сходства**: Сходство между запросом и документами оценивается с использованием метрик семантического сходства (косинусное сходство, евклидово расстояние и т.д.).
4. **Ранжирование результатов**: Документы сортируются по степени релевантности к запросу.

#### Пример

Рассмотрим пример, в котором мы используем модель BERT для улучшения поиска по запросу.

1. **Запрос пользователя**: "лучшие рестораны в Москве"
2. **Документ**: "Топ 10 ресторанов в Москве: наши рекомендации"

##### Шаг 1: Предварительная обработка данных

- Токенизация: ["лучшие", "рестораны", "в", "Москве"]
- Лемматизация: ["лучший", "ресторан", "в", "Москва"]

##### Шаг 2: Преобразование в векторы

Используем модель BERT для получения эмбеддингов запроса и документа.



In [ ]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

query = "лучшие рестораны в Москве"
document = "Топ 10 ресторанов в Москве: наши рекомендации"

query_embedding = model.encode(query)
document_embedding = model.encode(document)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


##### Шаг 3: Оценка сходства

Вычисляем косинусное сходство между вектором запроса и документа.



In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

similarity = cosine_similarity([query_embedding], [document_embedding])
print(similarity)

[[0.8871364]]


##### Шаг 4: Ранжирование результатов

На основе значения сходства (например, 0.85) документ ранжируется как релевантный к запросу.

### 2. Рекомендательные системы

Рекомендательные системы используют семантический анализ для создания персонализированных рекомендаций, сравнивая предпочтения пользователей с векторами тем.

#### Принципы работы

1. **Сбор данных**: Сбор информации о предпочтениях пользователей (история просмотров, оценки, клики и т.д.).
2. **Преобразование в векторы**: Преобразование данных о пользователях и элементах (товары, фильмы и т.д.) в векторные представления.
3. **Оценка сходства**: Оценка сходства между векторами предпочтений пользователей и векторами элементов.
4. **Формирование рекомендаций**: Генерация рекомендаций на основе полученных оценок сходства.

#### Пример

Рассмотрим пример создания персонализированных рекомендаций для пользователя в системе рекомендований фильмов.

1. **Пользователь**: Предпочитает драмы и триллеры.
2. **Фильм**: "Интерстеллар" — жанры: научная фантастика, драма.

##### Шаг 1: Сбор данных

- История просмотров пользователя: ["Титаник", "Семь", "Исчезнувшая"]
- Жанры фильмов: ["драма", "триллер", "драма"]

##### Шаг 2: Преобразование в векторы

Используем модель эмбеддинга для жанров, например Word2Vec, чтобы получить векторные представления жанров и фильмов.



In [ ]:
from gensim.models import Word2Vec

# Обучаем модель на жанрах фильмов
genres = [["драма"], ["триллер"], ["научная фантастика"], ["драма"]]
model = Word2Vec(genres, vector_size=10, min_count=1)

user_preferences = ["драма", "триллер"]
movie_genres = ["научная фантастика", "драма"]

user_vector = sum([model.wv[genre] for genre in user_preferences]) / len(user_preferences)
movie_vector = sum([model.wv[genre] for genre in movie_genres]) / len(movie_genres)

##### Шаг 3: Оценка сходства

Вычисляем косинусное сходство между вектором предпочтений пользователя и вектором фильма.




In [ ]:
similarity = cosine_similarity([user_vector], [movie_vector])
print(similarity)

[[0.8043248]]


Результат косинусного сходства равный 0.8043248 означает, что векторы предпочтений пользователя и фильма имеют высокое сходство. Косинусное сходство измеряет угол между двумя векторами в многомерном пространстве и показывает, насколько они близки друг к другу.

Чем ближе значение косинусного сходства к 1, тем более похожи векторы. В вашем случае значение 0.8043248 близко к 1, что указывает на высокую степень сходства между предпочтениями пользователя и фильмом.



##### Шаг 4: Формирование рекомендаций

На основе значения сходства (например, 0.80) фильм "Интерстеллар" может быть рекомендован пользователю.

Таким образом, семантический анализ на основе сходства векторов тем значительно улучшает качество информационного поиска и персонализации в рекомендательных системах. Использование моделей эмбеддинга и метрик семантического сходства позволяет более точно понимать и учитывать контекст и предпочтения пользователей, что ведет к более релевантным и полезным результатам.

#### 3. Масштабируемый семантический анализ и семантический поиск для больших корпусов текстов


## 3.1 Масштабируемость моделей

Обработка больших объемов текстовых данных требует высокоэффективных алгоритмов и мощных распределенных систем. Чтобы обеспечить масштабируемость, необходимо использовать параллельные вычисления и распределенные хранилища данных, такие как Hadoop и Spark. Эти технологии позволяют обрабатывать данные, разделяя задачи на более мелкие части и распределяя их по множеству узлов.

### Параллельные вычисления

Параллельные вычисления включают выполнение нескольких вычислительных задач одновременно. Это возможно благодаря разделению данных и задач на более мелкие части, которые могут обрабатываться параллельно.

#### Пример

Предположим, у нас есть большой корпус текстов, состоящий из миллиардов документов. Обработка всех документов с использованием одной машины займет огромное количество времени. Однако, разбив этот корпус на меньшие части и распределив их между несколькими машинами, можно значительно ускорить процесс.

### Распределенные хранилища данных

Распределенные хранилища данных, такие как Hadoop Distributed File System (HDFS) и Amazon S3, позволяют хранить и обрабатывать данные в распределенной среде. Эти системы обеспечивают высокую доступность и отказоустойчивость данных.

#### Пример

Система Hadoop может использовать HDFS для хранения огромных объемов данных, обеспечивая параллельный доступ к этим данным. Spark, в свою очередь, может использовать HDFS как источник данных для выполнения распределенных вычислений.

## 3.2 Инструменты и технологии

Для реализации масштабируемого семантического анализа и поиска применяются различные инструменты и технологии. Рассмотрим два ключевых инструмента: Apache Spark и Distributed Machine Learning.

### 1. Apache Spark

Apache Spark — это платформа для распределенной обработки данных, обеспечивающая быстрый и масштабируемый анализ. Она поддерживает различные операции, такие как мапредьюс, и позволяет обрабатывать данные в памяти, что значительно ускоряет вычисления.

#### Основные компоненты Spark

- **Spark Core**: Основной модуль, обеспечивающий распределенные вычисления.
- **Spark SQL**: Модуль для работы с данными в формате SQL.
- **Spark Streaming**: Модуль для обработки потоковых данных.
- **MLlib**: Библиотека машинного обучения.
- **GraphX**: Библиотека для графовых вычислений.

#### Пример использования Spark для семантического анализа

Предположим, у нас есть большой корпус текстов, и мы хотим выполнить семантический анализ с использованием модели Word2Vec. Мы можем использовать Spark для распределенной обработки данных и обучения модели.

```python
from pyspark.sql import SparkSession
from pyspark.ml.feature import Word2Vec

# Создаем сессию Spark
spark = SparkSession.builder.appName("Word2VecExample").getOrCreate()

# Загружаем данные
data = spark.read.text("hdfs://path/to/corpus")

# Преобразуем данные в формат для Word2Vec
documentDF = data.rdd.map(lambda row: row.value.split(" ")).toDF(["text"])

# Обучаем модель Word2Vec
word2Vec = Word2Vec(vectorSize=100, minCount=0, inputCol="text", outputCol="result")
model = word2Vec.fit(documentDF)

# Преобразуем документы в векторы
result = model.transform(documentDF)
result.show()
```



### 2. Distributed Machine Learning

Для обработки больших данных используются распределенные версии алгоритмов машинного обучения. Например, для обучения моделей Word2Vec и GloVe можно использовать их распределенные реализации.

#### Распределенные версии Word2Vec и GloVe

- **Distributed Word2Vec**: Обучение модели Word2Vec с использованием параллельных вычислений на нескольких узлах.
- **Distributed GloVe**: Обучение модели GloVe на больших объемах данных с использованием распределенной среды.

#### Пример использования распределенного Word2Vec

В Apache Spark можно использовать распределенные вычисления для обучения модели Word2Vec на большом корпусе текстов.





```
from pyspark.ml.feature import Word2Vec

# Настройки модели
word2Vec = Word2Vec(vectorSize=100, minCount=0, inputCol="text", outputCol="result")

# Обучение модели
model = word2Vec.fit(documentDF)

# Преобразование документов в векторы
result = model.transform(documentDF)
result.show()

```



### Пример использования распределенного GloVe

GloVe может быть реализован в распределенной среде с использованием подходов, аналогичных тем, что используются в Word2Vec. Существуют специализированные библиотеки, такие как Spark GloVe, которые позволяют обучать GloVe на больших объемах данных.

Таким образом, масштабируемый семантический анализ и поиск требуют использования эффективных алгоритмов и распределенных систем. Apache Spark и распределенные версии алгоритмов машинного обучения, таких как Word2Vec и GloVe, предоставляют мощные инструменты для обработки больших корпусов текстов. Использование параллельных вычислений и распределенных хранилищ данных позволяет значительно ускорить процесс и улучшить качество анализа.


**3.3 Обработка больших данных**

Для работы с большими корпусами текстов применяются следующие подходы:
1. **MapReduce**: Парадигма для обработки больших данных, разделяющая задачи на подзадачи и выполняющая их параллельно.
2. **Batch Processing и Stream Processing**: Batch Processing для анализа больших объемов данных за раз и Stream Processing для обработки данных в реальном времени.



#### 4. Семантические компоненты (темы) как признаки в конвейере NLP

**4.1 Интеграция семантических компонентов**

Векторные представления тем могут быть использованы в качестве признаков в различных NLP-задачах, таких как классификация текстов, кластеризация, распознавание сущностей и др.

**4.2 Примеры применения**

1. **Классификация текстов**: Использование векторов тем для обучения моделей классификации, например, для определения тематики документа.
2. **Кластеризация документов**: Группировка документов по семантической схожести с использованием векторных представлений.



**4.3 Построение конвейеров NLP**

1. **Feature Extraction**: Извлечение векторных признаков из текста.
2. **Model Training**: Обучение моделей на основе этих признаков.
3. **Model Evaluation**: Оценка качества моделей с использованием метрик, таких как точность, полнота и F-мера.



#### 5. Ориентация в векторных пространствах высокой размерности

**5.1 Высокая размерность и проблема курсовых**

Проблема курсовых возникает при работе в пространствах высокой размерности, где расстояния между точками становятся менее информативными.

**5.2 Методы уменьшения размерности**

Для работы в высокоразмерных пространствах часто используются методы уменьшения размерности:
1. **Principal Component Analysis (PCA)**: Преобразует данные в меньшее число измерений, сохраняя как можно больше вариации данных.
2. **t-SNE**: Метод для визуализации высокоразмерных данных, сохраняющий локальную структуру данных.
3. **UMAP**: Современный метод уменьшения размерности, который может быть более эффективным для некоторых задач, чем t-SNE.



**5.3 Практическое применение**

1. **Визуализация данных**: Использование методов уменьшения размерности для визуализации и анализа векторов слов и тем.
2. **Повышение эффективности алгоритмов**: Уменьшение размерности для ускорения работы алгоритмов машинного обучения и снижения потребления памяти.





Таким, образом, семантический анализ и векторное представление слов и тем являются фундаментальными компонентами современных систем NLP. Они позволяют создавать мощные и масштабируемые решения для поиска и анализа текстов. Понимание этих методов и их правильное применение открывает широкие возможности для разработки инновационных приложений и сервисов в области обработки естественного языка.

### Вопросы для закрепления темы

1. Что такое семантический анализ и какова его основная цель?
2. Объясните, что такое векторное представление слов (word embeddings) и его роль в семантическом анализе.
3. Назовите и кратко опишите три основные модели для создания векторных представлений слов.
4. Чем отличаются модели Word2Vec CBOW и Skip-Gram? Запишите их целевые функции.
5. Выведите формулу для вычисления вероятности $P(w_{t+j} | w_t)$ в моделях CBOW и Skip-Gram.
6. Как можно использовать векторные представления слов для поиска семантически близких слов? Опишите соответствующий алгоритм.
7. Предложите способ визуализации высокомерных векторных пространств слов. Какую информацию можно из этого извлечь?
8. Объясните, как можно применять векторные представления слов в задачах классификации текстов.
9. Разработайте методику для оценки качества векторных представлений, полученных с помощью различных моделей.
10. Опишите, как векторные представления слов могут использоваться для извлечения ключевых фраз из текстов.
11. Как можно использовать векторные представления слов для улучшения качества машинного перевода?
12. Объясните, как векторные представления слов могут применяться в системах вопросно-ответного диалога.
13. Предложите алгоритм для автоматического определения тем в тексте на основе анализа векторных представлений слов.
14. Как векторные представления слов могут использоваться для оценки сходства между текстами?
15. Опишите, как можно построить семантические кластеры слов, используя их векторные представления.
16. Объясните, как смещения (bias) в векторных представлениях слов могут быть выявлены и устранены.
17. Как можно обогатить векторные представления слов с помощью внешних источников знаний, таких как словари и онтологии?
18. Опишите, как векторные представления слов могут быть использованы для автоматического определения отношений между сущностями в тексте.
19. Предложите способ использования векторных представлений слов для автоматического порождения вопросов.
20. Объясните, как совместное обучение векторных представлений слов и их частей (морфем, префиксов, суффиксов) может улучшить качество семантического анализа.
21. Как можно применить векторные представления слов для автоматической классификации текстов на основе их семантической близости?
22. Опишите, как векторные представления слов могут использоваться для автоматического установления ссылок между текстами (text linking).
23. Объясните роль семантического анализа в задачах обработки естественного языка. Приведите примеры.
24. Сравните модели Word2Vec, GloVe и FastText. Какие преимущества и недостатки у каждой из них?
25. Выведите формулу для вычисления скалярного произведения в моделях CBOW и Skip-Gram. Объясните, как она используется при обучении.
26. Реализуйте простую версию модели CBOW на Python. Как можно обучить эту модель и проанализировать полученные векторные представления?
27. Разработайте алгоритм для поиска семантически близких слов в векторном пространстве. Как можно применить его для расширения поисковых запросов?
28. Предложите способ визуализации высокомерных векторных представлений слов. Какую информацию о семантике слов можно получить из такой визуализации?
29. Опишите, как векторные представления слов могут использоваться в задачах классификации текстов. Приведите пример реализации.
30. Объясните, как можно оценить качество векторных представлений, полученных с помощью различных моделей. Какие метрики можно использовать?

## Задачи для самостоятельной работы

1. Объясните основные принципы семантического анализа и векторного представления слов. Приведите примеры применения этих методов в реальных задачах обработки естественного языка.

2. Опишите различия между моделями Word2Vec CBOW и Skip-Gram. Сформулируйте целевые функции для каждой из этих моделей и поясните их интуитивный смысл.

3. Выведите формулы для вычисления вероятностей в моделях CBOW и Skip-Gram. Объясните, как эти вероятности используются для обучения векторных представлений слов.

4. Реализуйте простую версию модели CBOW на Python, используя библиотеку NumPy. Обучите модель на небольшом корпусе текстов и проанализируйте полученные векторные представления.

5. Сравните производительность моделей Word2Vec, GloVe и FastText на задаче аналогий слов. Объясните различия в подходах этих моделей к построению векторных представлений.

6. Разработайте алгоритм для поиска семантически близких слов в векторном пространстве. Опишите, как можно использовать этот алгоритм для задачи расширения запросов в поисковых системах.

7. Предложите способ визуализации векторных представлений слов высокой размерности. Объясните, какую информацию о семантике слов можно извлечь из такой визуализации.

8. Опишите, как можно использовать векторные представления слов в задачах классификации текстов. Приведите пример реализации такого подхода.

9. Разработайте методику для оценки качества векторных представлений, полученных различными моделями. Обсудите возможные метрики и способы их интерпретации.

10. Объясните, как можно использовать векторные представления слов для решения задачи извлечения ключевых фраз из текстов. Предложите алгоритм, основанный на этом подходе.

11. Рассмотрите применение векторных представлений слов в задаче машинного перевода. Объясните, как эта информация может быть использована для улучшения качества перевода.

12. Опишите, как можно использовать векторные представления слов для построения систем вопросо-ответного диалога. Приведите пример реализации такой системы.

13. Разработайте алгоритм для автоматического определения тем в тексте, основанный на анализе векторных представлений слов. Протестируйте его на наборе текстовых данных.

14. Предложите способ использования векторных представлений слов для решения задачи автореферирования текстов. Объясните, как можно выделять наиболее важные предложения на основе семантической близости.

15. Исследуйте, как можно применять векторные представления слов для задачи автоматической генерации текстов. Опишите, как эта информация может быть использована для улучшения качества генерируемых текстов.

16. Проанализируйте, как можно использовать векторные представления слов для повышения точности задач распознавания именованных сущностей. Предложите алгоритм, основанный на этом подходе.

17. Объясните, как можно применять векторные представления слов для решения задачи автоматического определения тональности текста (sentiment analysis). Приведите примеры реализации такого подхода.

18. Разработайте модель для совместного обучения векторных представлений слов и предложений. Опишите, как можно использовать эти представления в задачах, таких как поиск релевантных документов.

19. Исследуйте возможности использования векторных представлений слов для решения задачи автоматической классификации текстов по тематикам. Сравните различные подходы и оцените их эффективность.

20. Предложите способ применения векторных представлений слов для решения задачи обнаружения ложной информации (fake news detection). Объясните, как семантическая информация может помочь в этой задаче.

21. Опишите, как можно использовать векторные представления слов для решения задачи автоматического определения сходства между текстами. Приведите примеры практического применения такого подхода.

22. Разработайте алгоритм для построения семантических кластеров слов, основываясь на их векторных представлениях. Объясните, как можно использовать эту информацию для задач обработки естественного языка.

23. Исследуйте, как можно применять векторные представления слов для решения задачи машинного перевода с использованием нейронных сетей. Опишите, как семантическая информация может помочь в этом процессе.

24. Предложите методику для анализа смещений (bias) в векторных представлениях слов. Объясните, как можно использовать эту информацию для выявления и устранения предвзятости в моделях обработки естественного языка.

25. Разработайте алгоритм для обогащения векторных представлений слов с помощью внешних источников знаний, таких как словари и онтологии. Объясните, как это может повысить качество семантического анализа.

26. Исследуйте применение векторных представлений слов в задаче автоматического определения отношений между сущностями в тексте. Опишите, как эта информация может быть использована для построения знаний.

27. Предложите способ использования векторных представлений слов для решения задачи автоматического порождения вопросов. Объясните, как семантическая информация может помочь в этой задаче.

28. Разработайте модель для совместного обучения векторных представлений слов и их частей (морфем, префиксов, суффиксов). Объясните, как это может улучшить качество семантического анализа, особенно для редко встречающихся слов.

29. Исследуйте возможности использования векторных представлений слов для задачи автоматической классификации текстов на основе их семантической близости. Сравните этот подход с другими методами классификации.

30. Предложите способ применения векторных представлений слов для решения задачи автоматического установления ссылок между текстами (text linking). Объясните, как семантическая информация может помочь в этой задаче.